File: packages-logicn/logicn-devtools-benchmarks/benchmarks/arithmetic-threshold/bench.cpp
// arithmetic-threshold — C++ native
// Compile: g++ -O2 -o bench-arithmetic bench.cpp
#include <cstdint>
#include <chrono>
#include <iostream>
#include <cstring>
#include <cstdlib>

int main(int argc, char* argv[]) {
    long long threshold = 200000000000000LL;
    for (int i=1;i<argc-1;++i)
        if (!std::strcmp(argv[i],"--threshold")) threshold=std::atoll(argv[i+1]);
    using clk = std::chrono::steady_clock;
    auto t0=clk::now();
    long long total=0,idx=0,additions=0; uint32_t checksum=0u;
    while (total<=threshold) {
        total+=idx; idx++; additions++;
        total+=idx; idx++; additions++;
        uint32_t ui=(uint32_t)(idx&0xFFFFFFFFLL);
        checksum=(checksum^ui)*2654435761u+ui;
    }
    double elapsed=std::chrono::duration<double,std::milli>(clk::now()-t0).count();
    std::cout<<"{"
        <<"\"runtime\":\"cpp\","
        <<"\"benchmark\":\"arithmetic-threshold-v2\","
        <<"\"threshold\":"<<threshold<<","
        <<"\"additions\":"<<additions<<","
        <<"\"checksum\":"<<checksum<<","
        <<"\"elapsedMs\":"<<elapsed<<","
        <<"\"additionsPerSecond\":"<<(additions/(elapsed/1000.0))
        <<"}"<<std::endl;
}

File: packages-logicn/logicn-devtools-benchmarks/benchmarks/arithmetic-threshold/bench.rs
// arithmetic-threshold — Rust native
// Compile: rustc -O -o bench-arithmetic bench.rs
use std::time::Instant;
use std::env;

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut threshold: i64 = 200_000_000_000_000i64;
    let mut i = 1;
    while i+1<args.len() {
        if args[i]=="--threshold" { threshold=args[i+1].parse().unwrap_or(threshold); }
        i+=2;
    }
    let t0=Instant::now();
    let mut total:i64=0; let mut idx:i64=0; let mut additions:i64=0; let mut checksum:u32=0u32;
    while total<=threshold {
        total+=idx; idx+=1; additions+=1;
        total+=idx; idx+=1; additions+=1;
        let ui=idx as u32;
        checksum=(checksum^ui).wrapping_mul(2654435761u32).wrapping_add(ui);
    }
    let elapsed=t0.elapsed().as_secs_f64()*1000.0;
    println!("{{\"runtime\":\"rust\",\"benchmark\":\"arithmetic-threshold-v2\",\"threshold\":{},\"additions\":{},\"checksum\":{},\"elapsedMs\":{:.3},\"additionsPerSecond\":{:.2}}}",
        threshold,additions,checksum,elapsed,additions as f64/(elapsed/1000.0));
}
